### PR TITLE
Add Resolve Comment feature

### LINKED ISSUE #1056: Resolving comments
With long threads on pull requests and issues, there's often feedback that you'd want resolve as either taken care of or no longer relevant.

GitHub will already collapse comments on diffs that have since been updated:

<div align="center">
screen shot 2018-01-31 at 10 20 16 am
</div>

It would be awesome if we could mark any comment as "resolved" in order to clean up threads and make them easier to review later

<div align="center">
<h6>New "resolve" button in dropdown menu</h6> 
screen shot 2018-01-31 at 10 22 54 am
<br>
<h6>A resolved comment</h6> 
screen shot 2018-01-31 at 10 34 44 am
</div>

In order to do this without an API from GitHub, I figure a convention could be used in the comment body.

```
Could you just add some tests?

#resolved
```

Refined GitHub could automatically collapse comments with `#resolved` in the body, and add a button to automatically edit the comment to tack on `#resolved` at the end.

Thoughts?

### PR DESCRIPTION
**Pull request by jamiebuilds**
Started on #1056, if anyone wants to take it over just go right ahead

New menu item on comments:

screen shot 2018-02-15 at 7 01 22 am

Collapsed comment when matched "Comment resolved by @username":

screen shot 2018-02-15 at 7 01 07 am

Expanded:

screen shot 2018-02-15 at 7 01 12 am

Closed #1056.

### CODE REVIEW

**Comment by jamiebuilds** [comment]
Things to do:

- Turn "resolve" button into "unresolve" button when comment is already resolved
- "edited by" could be replaced with "resolved by"
- I'm not sure if there are any cases where the user can't edit the comment by the actions dropdown exists
- This only works with thread comments, not code comments
- I don't watch the dom properly for updates, navigating around sometimes creates additional "resolve" buttons to be added

File: source/features/resolve-comments.js [hunk 1]
```diff
@@ -0,0 +1,74 @@
+import select from 'select-dom';
+import {h} from 'dom-chef';
+import {getUsername} from '../libs/utils';
+import * as icons from '../libs/icons';
+
+const RESOLVED_TEXT_REGEX = /^Comment resolved by @([a-zA-Z0-9-]+)/;
```
Comment (sindresorhus) [inline source/features/resolve-comments.js]: `0-9` => `\d`

File: source/features/resolve-comments.js [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+
+export default () => {
+  for (const comment of select.all('.comment.timeline-comment')) {
+    const body = select('.comment-body', comment);
+    const isResolved = RESOLVED_TEXT_REGEX.test(body.innerText);
```
Comment (sindresorhus) [inline source/features/resolve-comments.js]: `.innerText` => `.textContent`

Comment (sindresorhus) [reply source/features/resolve-comments.js]: Needs your thoughts in https://github.com/sindresorhus/eslint-plugin-unicorn/issues/156

File: source/features/resolve-comments.js [hunk 3]
```diff
@@ -0,0 +1,74 @@
+import select from 'select-dom';
+import {h} from 'dom-chef';
```
Comment (bfred-it) [inline source/features/resolve-comments.js]: To line 1

File: source/features/resolve-comments.js [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+        <button
+          type="button"
+          class="btn-link"
+          onClick={() => {
+            if (commentContainer.style.display === 'none') {
+              commentContainer.style.display = '';
+              toggle.innerHTML = icons.fold().outerHTML;
+            } else {
+              commentContainer.style.display = 'none';
+              toggle.innerHTML = icons.unfold().outerHTML;
```
Comment (bfred-it) [inline source/features/resolve-comments.js]: No need to pass through HTML; use `toggle.firstChild.replaceWith(icons.unfold())` (or `firstElementChild`)

File: source/features/resolve-comments.js [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+
+  // select('.review-comment');
+};
+
+
+
+
+
+<button type="button" class="btn-link text-gray float-right f6 outdated-comment-label show-outdated-button js-details-target tooltipped tooltipped-w" aria-expanded="false" aria-label="Alt + click to expand/collapse all">
+  <svg aria-hidden="true" class="octicon octicon-unfold position-relative mr-1" height="16" version="1.1" viewBox="0 0 14 16" width="14"><path fill-rule="evenodd" d="M11.5 7.5L14 10c0 .55-.45 1-1 1H9v-1h3.5l-2-2h-7l-2 2H5v1H1c-.55 0-1-.45-1-1l2.5-2.5L0 5c0-.55.45-1 1-1h4v1H1.5l2 2h7l2-2H9V4h4c.55 0 1 .45 1 1l-2.5 2.5zM6 6h2V3h2L7 0 4 3h2v3zm2 3H6v3H4l3 3 3-3H8V9z"></path></svg>
```
Comment (bfred-it) [inline source/features/resolve-comments.js]: -> `icons.js`

File: source/features/resolve-comments.js [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+const RESOLVED_TEXT_REGEX = /^Comment resolved by @([a-zA-Z0-9-]+)/;
+
+export default () => {
+  for (const comment of select.all('.comment.timeline-comment')) {
+    const body = select('.comment-body', comment);
+    const isResolved = RESOLVED_TEXT_REGEX.test(body.innerText);
+
+    if (isResolved) {
+      const commentContainer = select('.edit-comment-hide', comment);
+      commentContainer.style.display = 'none';
```
Comment (bfred-it) [inline source/features/resolve-comments.js]: Perhaps we can try `commentContainer.hidden = true`

File: source/features/resolve-comments.js [hunk 7]
```diff
@@ -0,0 +1,74 @@
+import select from 'select-dom';
+import {h} from 'dom-chef';
+import {getUsername} from '../libs/utils';
+import * as icons from '../libs/icons';
+
+const RESOLVED_TEXT_REGEX = /^Comment resolved by @([a-zA-Z0-9-]+)/;
+
+export default () => {
+  for (const comment of select.all('.comment.timeline-comment')) {
```
Comment (bfred-it) [inline source/features/resolve-comments.js]: What do you think about selecting the "Edited" element (so we have to parse fewer elements) and then using `edited.closest('.comment')` to get the `comment` variable?

**Comment by bfred-it** [comment]
Looking forward to seeing this!

It should probably also work when there's no dropdown

**Comment by jamiebuilds** [comment]
Sorry, I really want to get back to this. I just have a lot of stuff to work on right now

**Comment by ramlmn** [comment]
This PR is running late!

**Comment by sindresorhus** [comment]
@ramlmn Your comment is not contributing anything to the discussion.

**Comment by ramlmn** [comment]
> Your comment is not contributing anything to the discussion ;)

@sindresorhus Just random ping to get the discussion going!

**Comment by jamiebuilds** [comment]
@ramlmn In the past month, I've moved from Australia to the US, been sick for two weeks, had to go to the emergency room because I was having trouble breathing, and I started a new job. I'll get back to this PR when I have a chance, it's not hurting anyone sitting around in the meantime while I have more important things going on.

**Comment by yakov116** [comment]
@jamiebuilds Wish you a full recovery. We can wait for you to your personal things "Resolved" first!

**Comment by ramlmn** [comment]
Apologies if I was being rude! All I wanted to know was the status of this PR as it has been more than a month (now we know why).

And as @yakov116 mentioned, this can wait till your personal things get "Resolved"!

**Comment by bfred-it** [comment]
Boom! https://blog.github.com/2018-04-18-new-tools-for-open-source-maintainers/

**Pull request closed without merging**